[docs] Update testing guide (#21863)
mui-org/material-ui: docs/src/pages/guides/testing/testing.md

<p class="description">Write tests to prevent regressions and write better code.</p>

## Userspace

> We recommend testing your application without tying the tests too close to Material-UI.

Reviewer: ```suggestion
It's generally recommended to test your application without tying the tests too closely to Material-UI.
```

> This is how we test our components internally.

Reviewer: ```suggestion
This is how Material-UI components are tested internally.
```

> A library that has a first-class API for this approach is [`@testing-library/react`](https://testing-library.com/docs/react-testing-library/intro).

Reviewer: I wonder if this sentence should come at the end of the section, after the approach has been fully explained? The prior sentences only allude to what *not* to do...

Author: That makes more sense :+1: I moved it after the example.

> For example, when rendering a `TextField` your test should not need to query for the specific Material-UI instance of the `TextField` but rather for the `input`, or the `[role="textbox"]`.

Reviewer: ```suggestion
For example, when rendering a `TextField` your test should not need to query for the specific Material-UI instance of the `TextField` but rather for the `input`, or `[role="textbox"]`.
```

By not relying on the React component tree you make your test more robust against internal changes in Material-UI or, if you need snapshot testing, adding additional wrapper components such as context providers.

> Though we don't recommend snapshot testing.

Reviewer: ```suggestion
We don't recommend snapshot testing though.
```
